packages/wp-source/types.ts

```diff
@@ -425,6 +442,11 @@ interface WpSource extends Source {
   };
 }
 
+/**
+ * Packages used internally by wp-source.
+ */
+export type Packages = WpSource & Package;
```

Comment:
How would it be different from the current `Package` interface? I'm sorry, it's not clear to me.

Reply:
It would be something like this:

```ts
interface Frontity extends Package {
  state: {
    frontity: {
      url: string;
      name: string;
      initialLink: string;
      //...
    };
  };
}
```

That way you only use it when your package needs to know about those. Right now there's no actions or libraries in the core, but there could be in the future.

We would get rid of those types in `Package`.

And finally, if you need the Frontity state in your package, you do this:

```ts
export type Packages = MyPackage & WpSource & Frontity;
```

instead of this

```ts
export type Packages = MyPackage & WpSource & Package;
```

Reply:
I think is a good idea. :slightly_smiling_face: 

By the way, did we take any decision regarding how package types should be merged? Because the intersection operator (`&`) was not working with TypeScript 3.9, was it?

Reply:
@DAreRodz Yes, @luisherranz has proposed that we'll need a type utility to merge the types instead of using the `&` operator. 

I've tried to create it and I came up with:

```ts
export type AnyArray<T = any> = Array<T> | ReadonlyArray<T>;

type Head<T extends AnyArray> = T["length"] extends 0 ? never : T[0];
type Tail<T extends AnyArray> = T["length"] extends 0
  ? never
  : ((...t: T) => void) extends (first: any, ...rest: infer Rest) => void
  ? Rest
  : never;

type Merge<T extends AnyArray> = Omit<Head<T>, 'name'> & Merge<Tail<T>>
```

The problem is that TS does not allow circular references (at least in some of the types) so it doesn't work 😕

Reply:
Let's continue the conversation in https://community.frontity.org/t/typescript-use-two-different-types/1681/16